### PR TITLE
Simple theme svg2png

### PR DESCRIPTION
**Pull request by dalf**
## What does this PR do?

In the master branch, convert (imagemagick) creates `searx/static/themes/simple/img/favicon.png` from `src/brand/searxng-wordmark.svg`. Some versions of convert have issues to render the .svg file.

This PR replaces the usage of convert by [sharp](https://github.com/lovell/sharp) (a standalone lib, it requires C++ to compile the library).

A custom grunt task is created because [grunt-task](https://www.npmjs.com/package/grunt-sharp) can't be installed (most probably because it hasn't been updated since a long time).

The "[build] /static" has been made using this change (the png file is smaller, it can be even smaller using [optipng](https://www.npmjs.com/package/optipng))

## Why is this change important?

Reliable build steps

## How to test this PR locally?

* `make themes.simple` or `npm run build`
* check `searx/static/themes/simple/img/favicon.png` is not modified (so `make static.build.commit` won't include the favicon at the build).
* check the favicon is rendered as expected.

### CODE REVIEW

File: searx/static/themes/simple/gruntfile.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ module.exports = function(grunt) {
     watch: {
       scripts: {
         files: ['gruntfile.js', 'src/**'],
-        tasks: ['eslint', 'copy', 'concat', 'svg2jinja', 'uglify', 'image', 'less:development', 'less:production']
+        tasks: ['eslint', 'copy', 'concat', 'uglify', 'less:development', 'less:production', 'image', 'svg2png', 'svg2jinja']
```
Comment (return42) [inline searx/static/themes/simple/gruntfile.js]: nit: Reordering the items in the tasks list makes it hard to see what has been changed.  Are there any special reason to reorder the tasks?

File: searx/static/themes/simple/gruntfile.js [hunk 2]
```diff
@@ -240,10 +269,11 @@ module.exports = function(grunt) {
...
     'concat',
-    'svg2jinja',
     'uglify',
-    'image',
     'less:development',
-    'less:production'
+    'less:production',
+    'image',
+    'svg2png',
+    'svg2jinja',
```
Comment (return42) [inline searx/static/themes/simple/gruntfile.js:298]: see my comment above ..

> Reordering the items in the task list makes it hard to see what has been changed. Are there any special reason to reorder the tasks?

File: searx/static/themes/simple/gruntfile.js [hunk 3]
```diff
@@ -100,16 +100,6 @@ module.exports = function(grunt) {
...
-    image: {
-      svg4web: {
-        options: {
-          svgo: ['--config', 'svg4web.svgo.js']
-        },
-        files: {
-          '<%= _templates %>/__common__/searxng-wordmark.min.svg': '<%= _brand %>/searxng-wordmark.svg'
-        }
-      }
-    },
```
Comment (return42) [inline searx/static/themes/simple/gruntfile.js:112]: see my comment above ..

> Reordering the items in the task list makes it hard to see what has been changed. Are there any special reason to reorder the tasks?

File: searx/static/themes/simple/gruntfile.js [hunk 4]
```diff
@@ -222,6 +228,29 @@ module.exports = function(grunt) {
     grunt.log.ok(this.data.dest + " created");
   });
 
+  grunt.registerMultiTask('svg2png', 'Convert SVG to PNG', function() {
```
Comment (return42) [inline searx/static/themes/simple/gruntfile.js]: Wow, very cool hadn't thought it is such easy :+1:

**Review by mrpaulblack — APPROVED** [review]
LGTM

I tested with `make node.env` and `make themes.simple` and generated new favicon properly and droped it with `make static.build.restore`; also the image is rendered properly in my browser

**Comment by return42** [comment]
I am still in a review / please wait ...

File: searx/static/themes/simple/gruntfile.js [hunk 5]
```diff
@@ -222,6 +228,29 @@ module.exports = function(grunt) {
...
+    const sharp = require('sharp');
+    for(const pngFileName in this.data.files) {
+      const svgFileName = this.data.files[pngFileName];
```
Comment (return42) [inline searx/static/themes/simple/gruntfile.js]: I don't know what is common, but intuitive I would suggest to change "key" & "value" of the dictionary:  

    files[<src svg>]: "<dest png>" 
 
or even better a list of destinations:

    files[<src svg>]: ["<dest-1 png>", "<dest-2 png> ...] 

EDIT: done in 171f492c

File: searx/static/themes/simple/gruntfile.js [hunk 6]
```diff
@@ -222,6 +228,29 @@ module.exports = function(grunt) {
...
+          .png({
+            force: true,
+            compressionLevel: 9,
+            palette: true,
+          })
+          .toFile(pngFileName, function (err) {
+            if (err !== null) {
+              grunt.log.error(err);
+            }
+          });
```
Comment (return42) [inline searx/static/themes/simple/gruntfile.js]: the `-resize 64x64` is missed here. (fixed in 171f492c)

**Review by return42 — CHANGES_REQUESTED** [review]
> it requires C++ to compile the library

Not on a _common_ build host, [Ready-compiled sharp and libvips binaries are provided for use on the most common platforms ... An ~7MB tarball containing libvips and its most commonly used dependencies is downloaded via HTTPS and stored within node_modules/sharp/vendor during npm install.](https://sharp.pixelplumbing.com/install#prebuilt-binaries)

> The "[build] /static" has been made using this change (the png file is smaller, it can be even smaller using optipng)

I missed the resize to 64x64 .. 5096 Byte --> 2747 Byte

I added commit 171f492c on top of your branch ...

    [mod] improve grunt task: svg2png / add resize options

.. it is my first Node.js implementation, please take a thorough look at it.

And I had some nits about reordering the tasks / see below.

File: searx/static/themes/simple/gruntfile.js (hunk 5 above)
Comment (dalf) [reply searx/static/themes/simple/gruntfile.js]: For these packages picked randomly:
* https://www.npmjs.com/package/grunt-image
* https://www.npmjs.com/package/grunt-svgmin
* https://www.npmjs.com/package/grunt-plantuml

Each time the pattern is:
```
{ "dest": "src" }
```

File: searx/static/themes/simple/gruntfile.js (hunk 1 above)
Comment (dalf) [reply searx/static/themes/simple/gruntfile.js]: Sorry, the purpose is to put all image tasks at the end, where it is easy to list which image is processed.
I should have done that in another PR.

File: searx/static/themes/simple/gruntfile.js [hunk 7]
```diff
@@ -137,6 +127,30 @@ module.exports = function(grunt) {
...
+    svg2png: {
+      favicon: {
+        options: {
+          width: 64,
+          hight: 64
+        },
+        files: [
+          {
+            src: '<%= _brand %>/searxng-wordmark.svg',
+            dest: [ 'img/favicon.png' ]
```
Comment (dalf) [inline searx/static/themes/simple/gruntfile.js]: The pattern
```
        files: {
          'img/favicon.png': '<%= _brand %>/searxng-wordmark.svg'
        }
```
make it easy to read that `'<%= _brand %>/searxng-wordmark.svg'` is used in this task and the one just above.

File: searx/static/themes/simple/gruntfile.js (hunk 1 above)
Comment (return42) [reply searx/static/themes/simple/gruntfile.js]: > Sorry, the purpose is to put all image tasks at the end, where it is easy to list which image is processed. I should have done that in another PR.

No need for a new PR, I can split the one commit in two commits, if you want.

**Review by return42 — APPROVED** [review]
I separated ..

64228b69 [mod] gruntfile: do some polish (no functional change) 

from my side this PR is ready to merge / @dalf if don't see any issue more, please merge.

**Comment by dalf** [comment]
I have added a commit to change 171f492

* the `svg2png` task define the files in a similar way to the `grunt-image` task:
   * https://github.com/searxng/searxng/blob/2109cfdfe45c76b20d9dd086a10719ef334841f9/searx/static/themes/simple/gruntfile.js#L150
   * https://github.com/searxng/searxng/blob/2109cfdfe45c76b20d9dd086a10719ef334841f9/searx/static/themes/simple/gruntfile.js#L157
* I removed the resize step: the size of the favicon image is 16x16 pixels, but the Firefox and other browsers display often used websites on a new tab. Without a [manifest.json endpoint](https://developer.mozilla.org/en-US/docs/Web/Manifest), the browser uses the favicon. A size of 16x16 will look pixelated. The current look is okay, so I removed the resize step:
image
* on error, grunt stops the compilation (instead of displaying a warning).
* use `await` to simplify the code ( see https://github.com/1000ch/grunt-image/blob/master/tasks/image.js )

If this change is accepted, the commits 79866a4  and 171f492  can be merged.

**Comment by return42** [comment]
> * the `svg2png` task define the files in a similar way to the `grunt-image` task: ...

OK

> * I removed the resize step: ...

I vote to get options back, ...

> Without a [manifest.json endpoint](https://developer.mozilla.org/en-US/docs/Web/Manifest), ...

With resize options we can easily create sizes, needed for a [manifest.json endpoint](https://developer.mozilla.org/en-US/docs/Web/Manifest) I can implement in another PR.

If it is OK for you to have resize options, would you like to add resize options and merge it with the two other commits / otherwise I can do it .. 

----

[EDIT] 

- SVG favicons https://github.com/audreyfeldroy/favicon-cheat-sheet#svg-file
- Definitive edition of "How to Favicon in 2021"  https://dev.to/masakudamatsu/favicon-nightmare-how-to-maintain-sanity-3al7

**Comment by return42** [comment]
> I vote to get options back, ...

Resolution options are no longer needed, in PR https://github.com/searxng/searxng/pull/558 favicon is generated as SVG.

**Review by return42 — APPROVED** [review]
FYI: squashed the two commits into 59f4c792b and rebased on current master.

**Pull request merged**